File: src-tauri/src/ipc/runtime_contract.rs
use serde::{Deserialize, Serialize};

use super::registry::{IPC_BINDING_HASH, IPC_CONTRACT_VERSION};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct RuntimeContractInfo {
    pub app_version: String,
    pub ipc_contract_version: u32,
    pub binding_hash: String,
    pub capabilities: Vec<RuntimeCapability>,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum RuntimeCapability {
    RuntimeContract,
    DataRecovery,
    Settings,
    Stations,
    StationKeys,
    Collectors,
    Routing,
    Proxy,
    ChannelMonitoring,
    Pricing,
    ChangeEvents,
    Capture,
    TypedStreaming,
}

pub const RUNTIME_CONTRACT_TYPESCRIPT: &str = r#"export type RuntimeCapability =
  | "runtime_contract"
  | "data_recovery"
  | "settings"
  | "stations"
  | "station_keys"
  | "collectors"
  | "routing"
  | "proxy"
  | "channel_monitoring"
  | "pricing"
  | "change_events"
  | "capture"
  | "typed_streaming";

export type RuntimeContractInfo = {
  appVersion: string;
  ipcContractVersion: number;
  bindingHash: string;
  capabilities: RuntimeCapability[];
};"#;

pub fn current_runtime_contract() -> RuntimeContractInfo {
    RuntimeContractInfo {
        app_version: env!("CARGO_PKG_VERSION").to_string(),
        ipc_contract_version: IPC_CONTRACT_VERSION,
        binding_hash: IPC_BINDING_HASH.to_string(),
        capabilities: vec![
            RuntimeCapability::RuntimeContract,
            RuntimeCapability::DataRecovery,
            RuntimeCapability::Settings,
            RuntimeCapability::Stations,
            RuntimeCapability::StationKeys,
            RuntimeCapability::Collectors,
            RuntimeCapability::Routing,
            RuntimeCapability::Proxy,
            RuntimeCapability::ChannelMonitoring,
            RuntimeCapability::Pricing,
            RuntimeCapability::ChangeEvents,
            RuntimeCapability::Capture,
            RuntimeCapability::TypedStreaming,
        ],
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn runtime_contract_is_safe_and_uses_compiled_registry_identity() {
        let contract = current_runtime_contract();
        assert_eq!(contract.ipc_contract_version, IPC_CONTRACT_VERSION);
        assert_eq!(contract.binding_hash, IPC_BINDING_HASH);
        assert_eq!(contract.app_version, env!("CARGO_PKG_VERSION"));
        let serialized = serde_json::to_value(contract).expect("serialize contract");
        assert!(serialized["capabilities"].is_array());
        assert!(serialized["capabilities"]
            .as_array()
            .expect("capabilities")
            .iter()
            .all(|value| value.is_string()));
    }

    #[test]
    fn runtime_contract_does_not_include_runtime_or_secret_state() {
        let serialized = serde_json::to_string(&current_runtime_contract()).expect("serialize");
        assert!(!serialized.contains("sqlite"));
        assert!(!serialized.contains("token"));
        assert!(!serialized.contains("secret"));
        assert!(!serialized.contains("\\"));
    }
}

File: src/app/bootstrap/runtimeContract.ts
import {
  IPC_BINDING_HASH,
  IPC_CONTRACT_VERSION,
  type RuntimeCapability,
  type RuntimeContractInfo,
} from "@/lib/bridge/contract";

const MAX_PAYLOAD_BYTES = 8_192;
const MAX_CAPABILITIES = 64;
const CONTRACT_KEYS = ["appVersion", "ipcContractVersion", "bindingHash", "capabilities"] as const;
const CAPABILITIES = new Set<RuntimeCapability>([
  "runtime_contract",
  "data_recovery",
  "settings",
  "stations",
  "station_keys",
  "collectors",
  "routing",
  "proxy",
  "channel_monitoring",
  "pricing",
  "alerting",
  "capture",
  "typed_streaming",
]);

export type RuntimeContractFailure =
  | "invalid_payload"
  | "version_mismatch"
  | "hash_mismatch"
  | "unknown_capability"
  | "missing_capability";

export type RuntimeContractValidation =
  | { ok: true; contract: RuntimeContractInfo }
  | { ok: false; reason: RuntimeContractFailure };

export function validateRuntimeContract(value: unknown): RuntimeContractValidation {
  if (!isRecord(value) || serializedSize(value) > MAX_PAYLOAD_BYTES) return { ok: false, reason: "invalid_payload" };
  const keys = Object.keys(value);
  if (keys.length !== CONTRACT_KEYS.length || keys.some((key) => !CONTRACT_KEYS.includes(key as typeof CONTRACT_KEYS[number]))) {
    return { ok: false, reason: "invalid_payload" };
  }
  if (typeof value.appVersion !== "string" || !safeVersion(value.appVersion)) return { ok: false, reason: "invalid_payload" };
  if (!Number.isInteger(value.ipcContractVersion) || typeof value.bindingHash !== "string" || !Array.isArray(value.capabilities)) {
    return { ok: false, reason: "invalid_payload" };
  }
  if (value.ipcContractVersion !== IPC_CONTRACT_VERSION) return { ok: false, reason: "version_mismatch" };
  if (value.bindingHash !== IPC_BINDING_HASH) return { ok: false, reason: "hash_mismatch" };
  if (value.capabilities.length === 0 || value.capabilities.length > MAX_CAPABILITIES) return { ok: false, reason: "invalid_payload" };
  const capabilities = value.capabilities.filter((capability): capability is RuntimeCapability => typeof capability === "string");
  if (capabilities.length !== value.capabilities.length || capabilities.some((capability) => !CAPABILITIES.has(capability))) {
    return { ok: false, reason: "unknown_capability" };
  }
  if (!capabilities.includes("runtime_contract")) return { ok: false, reason: "missing_capability" };
  const unique = new Set(capabilities);
  if (unique.size !== capabilities.length) return { ok: false, reason: "invalid_payload" };
  return {
    ok: true,
    contract: {
      appVersion: value.appVersion,
      ipcContractVersion: value.ipcContractVersion,
      bindingHash: value.bindingHash,
      capabilities,
    },
  };
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function serializedSize(value: unknown): number {
  try {
    return new TextEncoder().encode(JSON.stringify(value)).byteLength;
  } catch {
    return MAX_PAYLOAD_BYTES + 1;
  }
}

function safeVersion(value: string): boolean {
  return value.length > 0 && value.length <= 128 && /^[0-9A-Za-z.+_-]+$/.test(value);
}
